bazel_codegen/lib/src/assets/asset_reader.dart

```diff
@@ -0,0 +1,131 @@
+import 'dart:async';
+import 'dart:convert';
+
+import 'package:build/build.dart';
+import 'package:path/path.dart' as path;
+
+import '../errors.dart';
+import 'asset_filter.dart';
+import 'file_system.dart';
+
+/// An [AssetReader] tailored to bazel.
+///
+/// This allows specifying file paths via [loadAssetsByPath]. This will try to
+/// determine the appropriate [AssetId] for each input and use the provided path
+/// for that [AssetId] if it is requested. This allows us to use files which are
+/// generated by earlier bazel rules.
+class BazelAssetReader implements AssetReader {
+  /// The path to the package we are currently processing.
+  final String packagePath;
+
+  final _assetCache = <AssetId, String>{};
+
+  /// The google-specific file system.
+  ///
+  /// Responsible for knowing where bazel stores source and generated files on
+  /// disk.
+  final BazelFileSystem _fileSystem;
+
+  /// A filter for which assets are allowed to be read.
+  final AssetFilter _assetFilter;
+
+  final Map<String, String> _packageMap;
+
+  int numAssetsReadFromDisk = 0;
+
+  BazelAssetReader._(
+      this.packagePath, Iterable<String> rootDirs, this._packageMap,
+      {AssetFilter assetFilter})
+      : _fileSystem = new BazelFileSystem('.', rootDirs),
+        _assetFilter = assetFilter;
+
+  factory BazelAssetReader(String packagePath, Iterable<String> rootDirs,
+      Map<String, String> packageMap,
+      {AssetFilter assetFilter}) {
+    if (packagePath.endsWith('/') || packagePath.endsWith('\\')) {
```

Comment:
I think these are always going to have `/` coming from skylark rules, even on windows.